python/pyspark/pandas/series.py

```diff
@@ -2168,6 +2169,66 @@ def _fillna(
             )
         )._psser_for(self._column_label)
 
+    def interpolate(
+        self, method: Optional[str] = None, limit: Optional[int] = None
+    ) -> Optional["Series"]:
+        return self._interpolate(method=method, limit=limit)
+
+    def _interpolate(
+        self,
+        method: Optional[str] = None,
+        limit: Optional[int] = None,
+    ) -> "Series":
+        if (method is not None) and (method not in ["linear"]):
+            raise ValueError("Expecting 'linear'.")
+        if (limit is not None) and (not limit > 0):
+            raise ValueError("limit must be > 0.")
+
+        if not self.spark.nullable and not isinstance(
+            self.spark.data_type, (FloatType, DoubleType)
+        ):
+            return self._psdf.copy()._psser_for(self._column_label)
+
+        scol = self.spark.column
+        sql_utils = SparkContext._active_spark_context._jvm.PythonSQLUtils
+        last_non_null = Column(sql_utils.lastNonNull(scol._jc))
+        null_index = Column(sql_utils.nullIndex(scol._jc))
+
+        window_forward = Window.orderBy(NATURAL_ORDER_COLUMN_NAME).rowsBetween(
```

Comment:
Oh, okay, I got why this is tricky now. It should know the previous value to compute the current value .. but this way will move all data into a single worker as partition specification cannot be provided. Hm, it will be expensive but I don't have other ideas either.

Reply:
Based on SQL, I had no idea either.

but actually, this computation can be further optimzed in a scaled way:

1, perform global sort by `NATURAL_ORDER_COLUMN_NAME`;
2, extra pass on the sorted dataframe: collect some statistics within each physical partition, in this case, it needs: `size of the first and last null sequence`, `the first and last non-null value`.
3, based on the statistics collected from each physical partition, we then compute the state at the begining of each partition, and so can eval the expressions on each rows.

This computation was quite similar to what the `roc` is computed in `BinaryClassificationMetrics`, see https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala#L178-L253  (1, `sortByKey(ascending = false)`; 2, compute the `partitionwiseCumulativeCounts`, 3, fetch the `cumCount` on each partition and compute the curve).


When implementing `ewm`, I also tried this method (see https://github.com/apache/spark/pull/36062/files).

Comparing with exisiting plan, it should be more scalable, but if the dataset is small enough, it maybe solwer since it requires an extra pass.

I think it maybe up to the SQL optimizer to support computing the `window expression with no partition` in a more scaled way, since it seems a generic pattern (`row_number`/`rank`/`dense_rank` in SQL, `autocorr`/`ewm`/`null_index`/`last_non_null` in Python, can all be computed in this way.).